shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/Server.java

```diff
@@ -2631,15 +2633,33 @@ public void downloadInputData(
       try (var os = t.getResponseBody()) {}
       return;
     }
-    t.getResponseHeaders().set("Content-type", "application/json");
-    t.sendResponseHeaders(200, 0);
+
+    // If the format opts to use ErrorableStream itself, OK may be true or false. Otherwise,
+    // constructor will automatically set OK to true
+    ErrorableStream<Object> fetchedInput =
+          new ErrorableStream<>(inputSource.fetch(format.name(), readStale));
+    if (fetchedInput.isOk()) {
+      t.getResponseHeaders().set("Content-type", "application/json");
+      t.sendResponseHeaders(200, 0);
+    } else {
+      t.sendResponseHeaders(503, 0);
+      try (var os = t.getResponseBody()) {}
+        /**
+         * FIXME
+         * Previously, writeJson included a call to fetch, now we download the input much earlier.
+         * Do i need to release the inputDownloadSemaphore for any failures here too?
```

Comment:
By "failures here" do you mean failure to get the response body, and releasing the inputDownloadSemaphore in a corresponding catch?

Reply:
by 'failures here' i mean this case where !fetchedInput.isOk and we're returning 503

Reply:
Currently I have it releasing the semaphore but in the debugger it looks like that makes the semaphore value climb beyond what it was initially set to and i'm not sure that's supposed to happen

Reply:
That debugger output is indeed weird. The `return` on line 2654 means the second `release` on line 2662 shouldn't be tripped, and I can't see how the `release` on line 2653 would be called twice. Is this a debugger running on a test, or have you deployed locally and attached the debugger to check out a more real-world query?

Reply:
I observed this with shesmu deployed locally and using the /input and /input/stale endpoints, which are what will be used in practice as well

Reply:
That does indeed sound like a weird bug, the semaphore counter shouldn't go beyond what it was initially set to.

Reply:
If you don't release the semaphore here, does the semaphore eventually run out of available slots?

Reply:
Looks like it. 
Testing more, this happens specifically when readStale is true because it never hits the tryAcquire at 2631.

Reply:
Ok, I'll keep an eye out for a commit that fixes this.